devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,343 @@
+            // auto nonConstThis = const_cast<ICubForceTorque6DSensor*>(this);
+            // nonConstThis->setStatus(wearable::sensor::SensorStatus::Error);
+        }
+
+        return true;
+    }
+};
+
+// ==========================================
+// ICub constructor/destructor implementation
+// ==========================================
+ICub::ICub()
+    : pImpl{new ICubImpl()}
+{}
+
+ICub::~ICub() = default;
+
+// ======================
+// DeviceDriver interface
+// ======================
+bool ICub::open(yarp::os::Searchable& config)
+{
+    yInfo() << LogPrefix << "Starting to configure";
+
+    // Configure clock
+    if (!yarp::os::Time::isSystemClock()) {
+        yarp::os::Time::useSystemClock();
+    }
+
+    // ===============================
+    // CHECK THE CONFIGURATION OPTIONS
+    // ===============================
+
+    yarp::os::Bottle wbdHandsFTSensorsGroup = config.findGroup("wbd-hand-ft-sensors");
+    if (wbdHandsFTSensorsGroup.isNull()) {
+        yError() << LogPrefix << "REQUIRED parameter <wbd-hand-ft-sensors> NOT found";
+        return false;
+    }
+
+    if (!wbdHandsFTSensorsGroup.check("leftHand") || !wbdHandsFTSensorsGroup.check("rightHand")) {
+        yError() << LogPrefix << "REQUIRED parameter <leftHand> or <rightHand> NOT found";
+        return false;
+    }
+
+    // ===============================
+    // PARSE THE CONFIGURATION OPTIONS
+    // ===============================
+
+    pImpl->nSensors = wbdHandsFTSensorsGroup.size() - 1;
+    std::string leftHandFTPortName =
+        wbdHandsFTSensorsGroup.check("leftHand", yarp::os::Value(false)).asString();
+    std::string rightHandFTPortName =
+        wbdHandsFTSensorsGroup.check("rightHand", yarp::os::Value(false)).asString();
+
+    // ===========================
+    // OPEN AND CONNECT YARP PORTS
+    // ===========================
+
+    if (!(pImpl->leftHandFTPort.open("/ICub/leftHantFTSensor:i")
+          && yarp::os::Network::connect(leftHandFTPortName,
+                                        pImpl->leftHandFTPort.getName().c_str()))) {
+        yError() << LogPrefix << "Failed to open or connect to "
+                 << pImpl->leftHandFTPort.getName().c_str();
+        return false;
+    }
+
+    pImpl->sensorNames.push_back("leftWBDFTSensor");
+
+    if (!(pImpl->rightHandFTPort.open("/ICub/rightHantFTSensor:i")
+          && yarp::os::Network::connect(rightHandFTPortName,
+                                        pImpl->rightHandFTPort.getName().c_str()))) {
+        yError() << LogPrefix << "Failed to open or connect to "
+                 << pImpl->rightHandFTPort.getName().c_str();
+        return false;
+    }
+
+    pImpl->sensorNames.push_back("rightWBDFTSensor");
+
+    // Enable the callback of the ports
+    pImpl->leftHandFTPort.useCallback();
+    pImpl->rightHandFTPort.useCallback();
+
+    // Wait to receive first data
+    while (pImpl->leftHandFTPort.firstRun && pImpl->rightHandFTPort.firstRun) {
```

Comment:
This should wait that both ports received the first data. With `&&` the flow continues when just one did. Using `||` would solve it.

Reply:
@Yeshasvitvs Still TODO

Reply:
This comment seems to be outdated.

Reply:
If the code changed, we can resolve this conversation